src/Sentry/GlobalSessionManager.cs

```diff
@@ -0,0 +1,218 @@
+                    nic.NetworkInterfaceType != NetworkInterfaceType.Loopback)
+                .Select(nic => nic.GetPhysicalAddress().ToString())
+                .FirstOrDefault();
+
+            if (string.IsNullOrWhiteSpace(installationId))
+            {
+                _options.DiagnosticLogger?.LogError(
+                    "Failed to resolve hardware installation ID."
+                );
+
+                return null;
+            }
+
+            return installationId;
+        }
+
+        private string? TryGetInstallationId()
+        {
+            // Installation ID could have already been resolved by this point
+            if (!string.IsNullOrWhiteSpace(_cachedInstallationId))
+            {
+                return _cachedInstallationId;
+            }
+
+            var id =
+                TryGetPersistentInstallationId() ??
+                TryGetHardwareInstallationId();
+
+            if (!string.IsNullOrWhiteSpace(id))
+            {
+                _options.DiagnosticLogger?.LogDebug(
+                    "Resolved installation ID '{0}'.",
+                    id
+                );
+            }
+            else
+            {
+                _options.DiagnosticLogger?.LogDebug(
+                    "Failed to resolve installation ID."
+                );
+            }
+
+            return _cachedInstallationId = id;
+        }
+
+        public Session? StartSession()
+        {
+            // Extract release
+            var release = ReleaseLocator.Resolve(_options);
+            if (string.IsNullOrWhiteSpace(release))
+            {
+                // Release health without release is just health (useless)
+                _options.DiagnosticLogger?.LogError(
+                    "Failed to start a session because there is no release information."
+                );
+
+                return null;
+            }
+
+            // Extract other parameters
+            var environment = EnvironmentLocator.Resolve(_options);
+            var distinctId = TryGetInstallationId();
+
+            // Create new session
+            var session = new Session(distinctId, release, environment);
+
+            // Set new session and check whether we ended up overwriting an active one in the process
+            var previousSession = Interlocked.Exchange(ref _currentSession, session);
+            if (previousSession is not null)
+            {
+                _options.DiagnosticLogger?.LogWarning(
+                    "Starting a new session while an existing one is still active."
+                );
+
+                // End previous session
+                EndSession(previousSession, SessionEndStatus.Exited);
+            }
+
+            _options.DiagnosticLogger?.LogInfo(
+                "Started new session (SID: {0}; DID: {1}).",
+                session.Id, session.DistinctId
+            );
+
+            return session;
+        }
+
+        public void ReportError()
```

Comment:
I just learned that errors are no longer a counter. It's either 0 or 1.
It'll work as it is, but we can simplify things but removing it.

Reply:
Really? Even on client mode? Was this a recent change, because it's still a number in docs.